proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1097 @@
+   and would be the first instance of a lower-case type variable being in scope at the top level.
+
+``let`` in patterns
+~~~~~~~~~~~~~~~~~~~
+
+This part of this proposal allows introducing a ``let``\ -binding in a pattern.
+The bound variable(s) scope over the same region of code as the pattern-bound
+variables do.
+
+The syntax for this feature is a bit awkward, and so this proposal presents
+two alternatives for discussion.
+
+Motivation
+^^^^^^^^^^
+
+1. A careful reader will note that allowing ``let`` for `type synonyms <#type-let>`_
+   does not, by itself, replace a binding such as that in ``f (True :: b) = ...``
+   because the current form binds the type variable in the pattern. This part
+   of the proposal closes this gap. See `examples <#let-in-pattern-example>`_ below.
+
+#. Though admittedly a weakish motivation, there is currently no way to share
+   expressions used in common in multiple view patterns. See `examples <#let-in-pattern-example>`_
+   below.
+
+Specification Alternative 1
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+1. With ``-XExtendedLet``, add a new form of pattern as follows::
+
+     pat → 'let' decls 'in' pat
+
+#. Any entites bound in ``decls`` scope over the same region of the program
+   that pattern-bound variables scope over, with the addition of the ``decls``
+   themselves (that is, the declarations can be recursive).
+
+Specification Alternative 2
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+0. **Background**. Here are some productions from the `Haskell 2010 Report`_::
+
+     funlhs → var apat {apat}
+            | pat varop pat
+            | '(' funlhs ')' apat {apat}
+
+     apat → gcon
+          | literal
+          | …
+
+     lexp → '\' apat1 … apatn '->' exp   (n ≧ 1)
+          | …
+
+     lpat → apat
+          | '-' (integer|float)
+          | gcon apat1 … apatn
+
+   (Recall that braces mean "0 or more".)
+
+1. With ``-XExtendedLet``, change the grammar to be ::
+
+     funlhs → var apats1
+            | pat varop pat
+            | '(' funlhs ')' apats
+
+     apats1 → apat
+            | apat apats
+            | '(' 'let' decls ')' apats
+
+     apats →
+           | apats1
+
+     lexp → '\' apats1 '->' exp
+          | …
+
+     lpat → apat
+          | '-' (integer|float)
+          | gcon apats
+
+   This allows phrases like ``f x (let y = g x x) (frob y -> True) = ...``, where we can include
+   a ``let`` construct in the middle of a list of patterns. The pattern grammar itself is unaffected.
+
+2. Any entities bound in ``decls`` scope over the same region of the program
+   that pattern-bound variables scope over, with the addition of the ``decls``
+   themselves (that is, the declarations can be recursive).
+
+Examples
+^^^^^^^^
+
+.. _let-in-pattern-example:
+
+1. Instead of ::
+
+     f :: Maybe Bool -> Bool -> Bool
+     f (x :: Maybe b) (y :: b) = ...
+
+   we can write (Alternative 1) ::
+
+     f :: Maybe Bool -> Bool -> Bool
+     f (let type b = Bool in x) (y :: b) = ...
```

Comment:
I've got to say: this doesn't look very reasonable.

Reply:
A little history: I first wanted only Alternative 2, but someone in casual conversation found that quite strange (a pattern that doesn't match anything) and suggested Alternative 1.

Which is to say: I generally agree with you.